### PR TITLE
Support `__rand__`, `__ror__` and `__rxor__`

### LINKED ISSUE #58120: Support `__rand__`, `__ror__` and `__rxor__`
## 🚀 Feature
Support `torch.Tensor.{__rand__, __ror__, __rxor__}`.
(cc: @mruberry, @rgommers, @emcastillo and @kmaehashi)

## Motivation and Pitch

To enhance the compatibility with NumPy’s interface (c.f. https://github.com/pytorch/pytorch/issues/38349).

```py
>>> 5 & numpy.array([1, 2, 3, 4, 5])
array([1, 0, 1, 4, 5])
>>> 5 & torch.tensor([1, 2, 3, 4, 5])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for &: 'int' and 'Tensor'
```

## References

* [Python Array API: Reflected Operators](https://data-apis.org/array-api/latest/API_specification/array_object.html#reflected-operators)

cc @mruberry @rgommers @heitorschueroff

### PR DESCRIPTION
**Pull request by asi1024**
Fixes #58120.

This PR implements `torch.Tensor.{__rand__/__ror__/__rxor__}` for the compatibility with NumPy’s interface.
(cc: @mruberry, @rgommers, @emcastillo, @kmaehashi)

### CODE REVIEW

**Comment by asi1024** [comment]
The CI failure (test_unsupported_dtypes_renorm_cpu_float16) seems unrelated to this PR.

**Comment by mruberry** [comment]
You're correct, @asi1024, the renorm failures were, unfortunately, in trunk.

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -5744,6 +5744,30 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
            assert_autodiffed=True,
            supports_forward_ad=True,
            autodiff_nonfusible_nodes=['aten::mul'],),
+    OpInfo('__rand__',
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5747]: These are really cool. Would you also add some tests in test_binary_ufuncs.py that compares these operators with their tensor x scalar results? That is, a test that ensures that scalar x tensor and tensor x scalar is equivalent for all these operations?

**Review by mruberry — COMMENTED** [review]
Hey @asi1024! Looks like this just needs a test in test_binary_ufunc.py. The test can programmatically enumerate the operators in question and ensure they have the same tensor x scalar behavior as scalar x tensor behavior. Please let me know if you have any questions. I've added myself as a reviewer.

**Comment by asi1024** [comment]
@mruberry Rebased to the current master branch and added tests for rop to `test_binary_ufunc.py`!

File: test/test_binary_ufuncs.py [hunk 2]
```diff
@@ -223,44 +223,56 @@ def test_comparison_ops_check_for_zerodim_tensor_overflow(self, device):
         with self.assertRaisesRegex(RuntimeError, 'value cannot be converted to type'):
             self.assertTrue(ts2 != t2)
 
-    # TODO: update to work on CUDA, too
-    @onlyCPU
     def test_bitwise_ops(self, device):
-        x = torch.randn(5, 5).gt(0)
-        y = torch.randn(5, 5).gt(0)
+        def check_op(x, y):
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:227]: I think there's a more elegant way to do this. What would you think about the following?

```
# Tests that the binary operators and, or, and xor (as well as their reflected and inplace versions) work properly (AKA &, ||, ^ and &=, |=, ^=)
@dtypes(torch.bool, torch.float32)
test_bitwise_ops(self, device, dtype):
  # Tensor x Tensor and Tensor x Scalar ops
  ops = (operator.and_, operator.iand, operator.or_, operator.ior, operator.xor, operator.ixor)
  inplace_ops = (operator.iand, operator.ior, operator.ixor)
  shapes = ((5,), (15, 15), (500, 500))

  for op, shape in product(ops, shapes):
    # Tests tensor x tensor case
    a = make_tensor(shape, device=device, dtype=dtype)
    b =  make_tensor(shape, device=device, dtype=dtype)
    a_np = a.cpu().clone().numpy()

    self.assertEqual(op(a, b), op(a_np, b_np))

    # Tests tensor x scalar case
    a = make_tensor(shape, device=device, dtype=dtype)
    b = make_tensor((), device='cpu', dtype=dtype).item()
    
    a_np = a.cpu().clone().numpy()

    self.assertEqual(op(a, b), op(a_np, b_np))
  
    # Tests scalar x tensor case (for ops which aren't inplace)
    if op in inplace_ops:
      continue

    self.assertEqual(op(b, a), op(b, a_np))
```

What would you think of an approach like this?

Comment (mruberry) [reply test/test_binary_ufuncs.py:227]: cc @pmeier as an fyi to consider how we can automate this in the future

**Review by mruberry — COMMENTED** [review]
This still looks good, @asi1024. I made a suggestion about a test alternative for you to review. I'm looking forward to hearing your thoughts on it.

File: test/test_binary_ufuncs.py (hunk 2 above)
Comment (asi1024) [reply test/test_binary_ufuncs.py]: It seems better than my implementation :smiley:. I updated the test code.

Comment (pmeier) [reply test/test_binary_ufuncs.py]: > fyi to consider how we can automate this in the future

Not sure how we can improve on this. One thing that I was thinking on for quite a bit is a tagging system for `OpInfo`'s. Apart from the major properties (unary, binary, ...) each `OpInfo` could have a set of enums that give some additional information. If we had something like that, we could simply do `@ops([op for op in op_db if "bitwise" in op.tags])`

Comment (mruberry) [reply test/test_binary_ufuncs.py]: Sure; typically we derive filters from the metadata. Instead of tags we could probably just add another attribute

**Comment by asi1024** [comment]
@mruberry I updated the tests! Could you take another look?

**Review by mruberry — APPROVED** [review]
Awesome! Thank you @asi1024. Sorry for the delay. 

Would just rebase this to address the merge conflicts and I'll merge it for you?

**Comment by asi1024** [comment]
@mruberry Rebased and resolved merge conflict of `test/test_binary_ufuncs.py`!

**Pull request closed without merging**